README.rst
==========
SOSProject
==========

Project was to design a basic website using real-time information from Dublin Bikes
Completed site can be found here http://ec2-34-217-47-55.us-west-2.compute.amazonaws.com/

Note
====

This project has been set up using PyScaffold 3.0.1. For details and usage
information on PyScaffold see http://pyscaffold.org/.


CHANGELOG.rst
=========
Changelog
=========

Sprint 1 (Scrum Leader: Shane)
===========

Sprint 2 (Scrum Leader: Sean)
===========

Sprint 2 (Scrum Leader: Oz)
===========


tests/test_skeleton.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import pytest
from sosproject.skeleton import fib

__author__ = "SBean92"
__copyright__ = "SBean92"
__license__ = "mit"


def test_fib():
    assert fib(1) == 1
    assert fib(2) == 1
    assert fib(7) == 13
    with pytest.raises(AssertionError):
        fib(-10)


SOSProject/tests/conftest.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
    Dummy conftest.py for sosproject.

    If you don't know what this is for, just leave it empty.
    Read more about conftest.py under:
    https://pytest.org/latest/plugins.html
"""
from __future__ import print_function, absolute_import, division

# import pytest


src/dbmodel/main.py
'''
Created on 19 Mar 2018

@author: shane, sean, oz
'''
#code adapted from https://www.youtube.com/watch?v=kRNXKzfYrPU

from flask import Flask
from flask_sqlalchemy import SQLAlchemy
from flask_marshmallow import Marshmallow

app = Flask(__name__)
app.config['SQLALCHEMY_DATABASE_URI'] = 'mysql://sos:ozflanagan1@sos-database.cvwfzmigbgkv.us-west-2.rds.amazonaws.com/sosdatabase'
app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False

db = SQLAlchemy(app)
ma = Marshmallow(app)

# Creates station location and id data
class StaticBikeData(db.Model):
    __tablename__ = 'bike_static'
    number = db.Column('number', db.Integer, primary_key = True)
    contract = db.Column('contract', db.Unicode)
    street = db.Column('street', db.Unicode)
    address = db.Column('address', db.Unicode)
    lat = db.Column('lat', db.Float)
    lng = db.Column('lng', db.Float)

# Creates station available bike and stands data 
class DynamicBikeData(db.Model):
    __tablename__ = 'bike'
    number = db.Column('number', db.Integer, primary_key = True)
    banking = db.Column('banking', db.Unicode)
    bonus = db.Column('bonus', db.Unicode)
    status = db.Column('status', db.Unicode)
    stands = db.Column('stands', db.Integer)
    a_stands = db.Column('a_stands', db.Integer)
    a_bikes = db.Column('a_bikes', db.Integer)
    timestamp = db.Column('timestamp', db.BigInteger, primary_key = True)

# Creates average bike and stands data by hour   
class AvgBikeData(db.Model):
    __tablename__ = 'averages_hourly'
    index = db.Column('index', db.Integer, primary_key = True)
    number = db.Column('number', db.Integer)
    weekday = db.Column('weekday', db.Unicode)
    hour = db.Column('hour', db.Integer)
    a_bikes = db.Column('a_stands', db.Integer)
    a_stands = db.Column('a_bikes', db.Integer)

# Creates daily weather data
class DailyWeatherData(db.Model):
    __tablename__ = 'daily_weather'
    timestamp = db.Column('timestamp', db.Unicode, primary_key = True)
    isRaining = db.Column('isRaining', db.Integer)
    description = db.Column('description', db.Unicode)

# Creates hourly weather data
class HourlyWeatherData(db.Model):
    __tablename__ = 'hourly_weather'
    timestamp = db.Column('timestamp', db.Unicode, primary_key = True)
    isRaining = db.Column('isRaining', db.Integer)
    description = db.Column('description', db.Unicode)

# Creates average bike and stands data by day
class AvgBikeDataDay(db.Model):
    __tablename__ = 'averages_daily'
    index = db.Column('index', db.Integer, primary_key = True)
    number = db.Column('number', db.Integer)
    weekday = db.Column('weekday', db.Unicode)
    a_bikes = db.Column('a_stands', db.Integer)
    a_stands = db.Column('a_bikes', db.Integer)

class StaticBikeSchema(ma.ModelSchema):
    class Meta:
        model = StaticBikeData
        
class DynamicBikeSchema(ma.ModelSchema):
    class Meta:
        model = DynamicBikeData
        
class AvgBikeSchema(ma.ModelSchema):
    class Meta:
        model = AvgBikeData

class DailyWeatherSchema(ma.ModelSchema):
    class Meta:
        model = DailyWeatherData

class HourlyWeatherSchema(ma.ModelSchema):
    class Meta:
        model = HourlyWeatherData
        
class AvgBikeDaySchema(ma.ModelSchema):
    class Meta:
        model = AvgBikeDataDay
        

src/sosproject/flask.py
'''
Created on 20 Mar 2018

@author: shane, sean, oz
'''
from flask import Flask, render_template
from flask_sqlalchemy import SQLAlchemy
from dbmodel.main import StaticBikeData, DynamicBikeData, StaticBikeSchema, DynamicBikeSchema, AvgBikeData, AvgBikeSchema, DailyWeatherData, DailyWeatherSchema, HourlyWeatherData, HourlyWeatherSchema,\
    AvgBikeDataDay, AvgBikeDaySchema
import pymysql
pymysql.install_as_MySQLdb()
from flask.json import jsonify

app = Flask(__name__)
app.config['SQLALCHEMY_DATABASE_URI'] = 'mysql://sos:ozflanagan1@sos-database.cvwfzmigbgkv.us-west-2.rds.amazonaws.com/sosdatabase'
app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = True
db = SQLAlchemy(app)

@app.route('/')

def index():
    return render_template('index.html')

@app.route('/root1')
# Root 1 contains station location and id data
def display1():
    static_bike_data = StaticBikeData.query.all()
    static_bike_schema = StaticBikeSchema(many=True)
    static_output = static_bike_schema.dump(static_bike_data).data
    return jsonify(static_output)

@app.route('/root2')
# Root 2 contains station available bike and stands data
def display2():
    dynamic_bike_data = DynamicBikeData.query.all()
    dynamic_bike_schema = DynamicBikeSchema(many=True)
    dynamic_output = dynamic_bike_schema.dump(dynamic_bike_data).data
    return jsonify(dynamic_output)

@app.route('/root3')
# Root 3 contains most recent station available bike and stands data
def display3():
    live_bike_data = db.engine.execute('SELECT number, LatestUpdate, banking, bonus, status, stands, a_stands, a_bikes FROM(SELECT * FROM bike INNER JOIN (SELECT MAX(timestamp) AS LatestUpdate, number AS number2 FROM bike GROUP BY number2) SubMax ON bike.timestamp = SubMax.LatestUpdate and bike.number = SubMax.number2)bikeLatest;')
    live_bike_schema = DynamicBikeSchema(many=True)
    live_output = live_bike_schema.dump(live_bike_data).data
    return jsonify(root3=live_output)

@app.route('/root4')
# Root 4 contains average bike and stands data by hour 
def display4():
    avg_bike_data = AvgBikeData.query.all()
    avg_bike_schema = AvgBikeSchema(many=True)
    avg_output = avg_bike_schema.dump(avg_bike_data).data
    return jsonify(root4=avg_output)

@app.route('/root5')
# Root 5 contains daily weather data
def display5():
    daily_weather_data = DailyWeatherData.query.all()
    daily_weather_schema = DailyWeatherSchema(many=True)
    daily_weather_output = daily_weather_schema.dump(daily_weather_data).data
    return jsonify(root5=daily_weather_output)

@app.route('/root6')
# Root 5 contains hourly weather data
def display6():
    hourly_weather_data = HourlyWeatherData.query.all()
    hourly_weather_schema = HourlyWeatherSchema(many=True)
    hourly_weather_output = hourly_weather_schema.dump(hourly_weather_data).data
    return jsonify(root6=hourly_weather_output)

@app.route('/root7')
# Root 7 contains average bike and stands data by day
def display7():
    avg_bike_data_day = AvgBikeDataDay.query.all()
    avg_bike_schema_day = AvgBikeDaySchema(many=True)
    avg_day_output = avg_bike_schema_day.dump(avg_bike_data_day).data
    return jsonify(root7=avg_day_output)

if __name__ == "__main__":
    app.run(debug=True)
    


src/sosproject/scraper.py
'''
Created on 16 Mar 2018

@author: shane, sean, oz

Practice for learning to access API
'''
from jsonschema.validators import requests
import csv
import time
import os
import MySQLdb
import traceback
from collections import OrderedDict

# Gets data from JCDecaux, writes to csv
def scraper():
    try:
        os.remove('bikes.csv')
        os.remove('static.csv')
    except OSError:
        pass
    
    response = requests.get("https://api.jcdecaux.com/vls/v1/stations?contract=Dublin&apiKey=cf7b17e2c126f2e1998430272919516c5acb5538")
    data = response.json(object_pairs_hook=OrderedDict)

    for i in data:
        myliststatic=[]
        mylistdynamic=[]
        for key, value in i.items():
            if key == "number":
                myliststatic.append(value)
                mylistdynamic.append(value)
            elif key == "position":
                for new_key, new_value in value.items():
                    myliststatic.append(new_value)
            elif key == "name" or key == "address" or key == "contract_name":
                myliststatic.append(value)
            else:
                mylistdynamic.append(value)
        
        with open(r"bikes.csv", 'a') as csvfile:
                bike_writer = csv.writer(csvfile, lineterminator = '\n')
                bike_writer.writerow(mylistdynamic)

        with open(r"static.csv", 'a') as csvfile:
                static_writer = csv.writer(csvfile, lineterminator = '\n')
                static_writer.writerow(myliststatic)

def stationCount():
    count = 0
    csv_data = csv.reader(open('bikes.csv', 'r'))
    for row in csv_data:
        count += 1
    return count

# Creates table on AWS database
def createTable(table):
    conn = MySQLdb.connect(host = 'sos-database.cvwfzmigbgkv.us-west-2.rds.amazonaws.com', user = 'sos', passwd = 'ozflanagan1', db = 'sosdatabase')
    cursor = conn.cursor()    
    if table == 'bike':
        cursor.execute('CREATE TABLE bike (number int(20) NOT NULL,street varchar(50),address varchar(50),lat float,lng float,banking varchar(5),bonus varchar(5),status varchar(10),contract varchar(10),stands int(3),a_stands int(3),a_bikes int(3),timestamp bigint(20) NOT NULL,CONSTRAINT PK_bike PRIMARY KEY (number,timestamp))')
        conn.commit()
    elif table == 'static':
        cursor.execute('CREATE TABLE bike_static(number int(4) NOT NULL, contract varchar(10), street varchar(50), address varchar(50), lat float, lng float, PRIMARY KEY (number),FOREIGN KEY (number) REFERENCES bike(number))')
        conn.commit()
    cursor.close()

def tableExist(table):
    conn = MySQLdb.connect(host = 'sos-database.cvwfzmigbgkv.us-west-2.rds.amazonaws.com', user = 'sos', passwd = 'ozflanagan1', db = 'sosdatabase')
    cursor = conn.cursor()
    cursor.execute("SHOW TABLES LIKE'"+table+"'")
    if cursor.fetchone():
        cursor.close()
        return True
    else:
        cursor.close()
        return False

# Writes to table on AWS database
def sqlWrite(table):
    conn = MySQLdb.connect(host = 'sos-database.cvwfzmigbgkv.us-west-2.rds.amazonaws.com', user = 'sos', passwd = 'ozflanagan1', db = 'sosdatabase')
    cursor = conn.cursor()
    if table == 'bike':
        lines = 0
        csv_data = csv.reader(open('bikes.csv', 'r'))
        for row in csv_data:
            lines += 1
            cursor.execute('INSERT INTO bike(number, banking, bonus, status, stands, a_stands, a_bikes, timestamp)' +
                            'VALUES(%s, %s, %s, %s, %s, %s, %s, %s)', row)
        conn.commit()
        print("Inserting",lines,"lines into dynamic table")
    if table == 'static':
        lines = 0
        csv_data = csv.reader(open('static.csv', 'r'))
        for row in csv_data:
            lines += 1
            cursor.execute('INSERT INTO bike_static(number, street, address, lat, lng, contract)' +
                            'VALUES(%s, %s, %s, %s, %s, %s)', row)
        conn.commit()
        print("Inserting",lines,"lines into static table")     
    cursor.close()    

def main():
    oldMaxStations = 0
    while True:
        if tableExist("bike"):
            pass
        else:
            createTable("bike")
        if tableExist("bike_static"):
            pass
        else:
            createTable("static")
        try:
            scraper()
        except Exception:
            print("Error in getting data")
            traceback.print_exc()
        try:
            sqlWrite("bike")
        except Exception:
            print("Error in inserting data into bike")
            traceback.print_exc()
        newMaxStations = stationCount()
        if oldMaxStations != newMaxStations:
            try:
                sqlWrite("static")
            except Exception:
                print("Error in inserting data into static")
                traceback.print_exc()
            oldMaxStations = newMaxStations
        time.sleep(600)

if __name__ == "__main__":
    main()

SOSProject/src/sosproject/skeleton.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
This is a skeleton file that can serve as a starting point for a Python
console script. To run this script uncomment the following line in the
entry_points section in setup.cfg:

    console_scripts =
     fibonacci = sosproject.skeleton:run

Then run `python setup.py install` which will install the command `fibonacci`
inside your current environment.
Besides console scripts, the header (i.e. until _logger...) of this file can
also be used as template for Python modules.

Note: This skeleton file can be safely removed if not needed!
"""
from __future__ import division, print_function, absolute_import

import argparse
import sys
import logging

from sosproject import __version__

__author__ = "SBean92"
__copyright__ = "SBean92"
__license__ = "mit"

_logger = logging.getLogger(__name__)


def fib(n):
    """Fibonacci example function

    Args:
      n (int): integer

    Returns:
      int: n-th Fibonacci number
    """
    assert n > 0
    a, b = 1, 1
    for i in range(n-1):
        a, b = b, a+b
    return a


def parse_args(args):
    """Parse command line parameters

    Args:
      args ([str]): command line parameters as list of strings

    Returns:
      :obj:`argparse.Namespace`: command line parameters namespace
    """
    parser = argparse.ArgumentParser(
        description="Just a Fibonnaci demonstration")
    parser.add_argument(
        '--version',
        action='version',
        version='SOSProject {ver}'.format(ver=__version__))
    parser.add_argument(
        dest="n",
        help="n-th Fibonacci number",
        type=int,
        metavar="INT")
    parser.add_argument(
        '-v',
        '--verbose',
        dest="loglevel",
        help="set loglevel to INFO",
        action='store_const',
        const=logging.INFO)
    parser.add_argument(
        '-vv',
        '--very-verbose',
        dest="loglevel",
        help="set loglevel to DEBUG",
        action='store_const',
        const=logging.DEBUG)
    return parser.parse_args(args)


def setup_logging(loglevel):
    """Setup basic logging

    Args:
      loglevel (int): minimum loglevel for emitting messages
    """
    logformat = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"
    logging.basicConfig(level=loglevel, stream=sys.stdout,
                        format=logformat, datefmt="%Y-%m-%d %H:%M:%S")


def main(args):
    """Main entry point allowing external calls

    Args:
      args ([str]): command line parameter list
    """
    args = parse_args(args)
    setup_logging(args.loglevel)
    _logger.debug("Starting crazy calculations...")
    print("The {}-th Fibonacci number is {}".format(args.n, fib(args.n)))
    _logger.info("Script ends here")


def run():
    """Entry point for console_scripts
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    run()


src/sosproject/weather_scraper.py
import requests
import json
import pymysql
import csv
import time
import os
import traceback

def daily_scraper():
    try:
        os.remove('daily_weather.csv')
    except:
        pass
    #Read in weather info and put into JSON Dump. Currently only doing it once
    #This is for the whole day and is a forecast. Will only keep current day then overwrite on next access
    try: 
        response = requests.get("http://api.openweathermap.org/data/2.5/forecast?id=2964574&APPID=4f2499afbefdfe83bed8ceeca402adc5")
        weather_forecast = response.json()
    except:
        print("Error connecting to API")
    """with open('weather_data.txt', 'w') as outfile:
        json.dump(weather, outfile, sort_keys = True, indent = 4,
                   ensure_ascii = False)
    
    
    #Just reading in from here so I don't get kicked off the API
    #this can be deleted as soon as it all works
    with open('weather_data.txt') as json_data:
        weather_forecast = json.load(json_data)"""
    
    
    #Regardless of whether current/ forecast this will be same
    #essentially taking in list in weather and seeing if any of rain codes present
    # if they are then have 1 value for rain, if not then 0 value for not raining
    
    weather_forecast= weather_forecast['list']
    rain_codes=[200,201,202,210,211,212,
                221,230,231,232,300,301,
                302,310,311,312,313,314,
                321,500,501,502,503,504,
                511,520,521,522,531,600,
                601,602,611,612,615,616,
                620,621,622]
    
    #Appends date, raining yes/no and overall description to list
    for i in weather_forecast:
        weather_list=[]
        weather_list.append(i['dt'])
        is_raining= (i['weather'][0]['id'])
        if is_raining in rain_codes:
            weather_list.append(1)
        else:
            weather_list.append(0)
        weather_list.append(i['weather'][0]['description'])
        print("daily",weather_list)
        
    #need a way to put list into DB for weather
         
        with open(r"daily_weather.csv", 'a') as csvfile:
            weather_writer = csv.writer(csvfile, lineterminator = '\n')
            weather_writer.writerow(weather_list)
            
def hourly_weather_scraper():
    try:
        os.remove('hourly_weather.csv')
    except OSError:
        pass
    #Read in weather info and put into JSON Dump. Currently only doing it once
    #This is current data, this will be constantly stored in the db
    try:
        response = requests.get("http://api.openweathermap.org/data/2.5/weather?id=2964574&APPID=4f2499afbefdfe83bed8ceeca402adc5")
        weather = response.json()
    except:
        print("Error connecting to API")
        
    """with open('data.txt', 'w') as outfile:
        json.dump(weather, outfile, sort_keys = True, indent = 4,
                   ensure_ascii = False)
    
    #Just reading in from here so I don't get kicked off the API
    with open('data.txt') as json_data:
        weather = json.load(json_data)"""
    
    
    #Regardless of whether current/ forecast this will be same
    #essentially taking in list in weather and seeing if any of rain codes present
    # if they are then have 1 value for rain, if not then 0 value for not raining
    rain_codes=[200,201,202,210,211,212,
                221,230,231,232,300,301,
                302,310,311,312,313,314,
                321,500,501,502,503,504,
                511,520,521,522,531,600,
                601,602,611,612,615,616,
                620,621,622]
    #Appends date, raining yes/no and overall description to list
    weather_list=[]
    weather_list.append(weather['dt'])
    if weather['weather'][0]['id'] in rain_codes:
        weather_list.append(1)
    else:
        weather_list.append(0)
    weather_list.append(weather['weather'][0]['description'])
    print("hourly",weather_list)
    with open(r"hourly_weather.csv", 'a') as csvfile:
        weather_writer = csv.writer(csvfile, lineterminator = '\n')
        weather_writer.writerow(weather_list)

def createWeatherTable(table):
    conn = pymysql.connect(host = 'sos-database.cvwfzmigbgkv.us-west-2.rds.amazonaws.com', user = 'sos', passwd = 'ozflanagan1', db = 'sosdatabase')
    cursor = conn.cursor()    
    if table == 'hourly':
        cursor.execute('CREATE TABLE hourly_weather (timestamp varchar(50) NOT NULL,isRaining int(1),description varchar(150),CONSTRAINT PK_hourly PRIMARY KEY (timestamp))')
        conn.commit()
    elif table == 'daily':
        cursor.execute('CREATE TABLE daily_weather(timestamp varchar(50) NOT NULL,isRaining int(1),description varchar(150),CONSTRAINT PK_daily PRIMARY KEY (timestamp))')
        conn.commit()
    cursor.close()

def sqlWriteWeather(table):
    try:
        conn = pymysql.connect(host = 'sos-database.cvwfzmigbgkv.us-west-2.rds.amazonaws.com', user = 'sos', passwd = 'ozflanagan1', db = 'sosdatabase')
    except:
        print("Problem connecting to the database")
    cursor = conn.cursor()
    if table == 'hourly':
        lines = 0
        csv_data = csv.reader(open('hourly_weather.csv', 'r'))
        for row in csv_data:
            lines += 1
            cursor.execute('INSERT INTO hourly_weather(timestamp, isRaining, description)' +
                            'VALUES(%s, %s, %s)', row)
        conn.commit()
        print("Inserting",lines,"lines into hourly weather table")
        
    if table == 'daily':
        lines = 0
        csv_data = csv.reader(open('daily_weather.csv', 'r'))
        for row in csv_data:
            lines += 1
            cursor.execute('INSERT INTO daily_weather(timestamp, isRaining, description)' +
                            'VALUES(%s, %s, %s)', row)
        conn.commit()
        print("Inserting",lines,"lines into daily weather table")     
    cursor.close()

def tableExist(table):
    conn = pymysql.connect(host = 'sos-database.cvwfzmigbgkv.us-west-2.rds.amazonaws.com', user = 'sos', passwd = 'ozflanagan1', db = 'sosdatabase')
    cursor = conn.cursor()
    cursor.execute("SHOW TABLES LIKE'"+table+"'")
    if cursor.fetchone():
        cursor.close()
        return True
    else:
        cursor.close()
        return False

def dropTable():
    conn = pymysql.connect(host = 'sos-database.cvwfzmigbgkv.us-west-2.rds.amazonaws.com', user = 'sos', passwd = 'ozflanagan1', db = 'sosdatabase')
    cursor = conn.cursor()
    cursor.execute("DROP TABLE daily_weather")
    cursor.close()

def main():
    while True:
        if tableExist("daily_weather"):
            dropTable()
        try:
            daily_scraper()
        except Exception:
            print("Error in getting daily data")
        createWeatherTable("daily")
        try:
            sqlWriteWeather("daily")
        except Exception:
            print("Error in inserting daily data")

        if tableExist("hourly_weather"):
            pass
        else:
            createWeatherTable("hourly")
        for x in range(0,47):
            try:
                hourly_weather_scraper()
            except Exception:
                print("Error in getting hourly data")
            try:
                sqlWriteWeather("hourly")
            except Exception:
                print("Error in inserting hourly data")
            time.sleep(1800)

if __name__ == "__main__":
    main()